pkg/daemon/update.go

```diff
@@ -382,29 +382,42 @@ func (dn *CoreOSDaemon) applyOSChanges(mcDiff machineConfigDiff, oldConfig, newC
 		dn.nodeWriter.Eventf(corev1.EventTypeNormal, "OSUpdateStarted", mcDiff.osChangesString())
 	}
 
+	// We previously did not emit this event when kargs changed, so we still don't
 	if mcDiff.osUpdate || mcDiff.extensions || mcDiff.kernelType {
 		// We emitted this event before, so keep it
 		if dn.nodeWriter != nil {
 			dn.nodeWriter.Eventf(corev1.EventTypeNormal, "InClusterUpgrade", fmt.Sprintf("Updating from oscontainer %s", newConfig.Spec.OSImageURL))
 		}
-
 	}
 
-	// The steps from here on are different depending on the image type, so check the image type
-	isLayeredImage, err := dn.NodeUpdaterClient.IsBootableImage(newConfig.Spec.OSImageURL)
-	if err != nil {
-		return fmt.Errorf("Error checking type of update image: %w", err)
-	}
+	// Only check the image type and excute OS changes if:
+	// - machineconfig changed
+	// - we're staying on a realtime kernel ( need to run rpm-ostree update )
+	// - we have extensions ( need to run rpm-ostree update )
+	// We have at least one customer that removes the pull secret from the cluster to "shrinkwrap" it for distribution and we want
+	// to make sure we don't break that use case, but realtime kernel update and extensions update always ran
+	// if they were in use, so we also need to preserve that behavior.
+	// https://issues.redhat.com/browse/OCPBUGS-4049
+	if mcDiff.osUpdate || mcDiff.extensions || mcDiff.kernelType || mcDiff.kargs ||
+		canonicalizeKernelType(newConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime || len(newConfig.Spec.Extensions) > 0 {
```

Comment:
The additional non-diff conditions are to allow realtime kernel and extensions to update their packages if they are in use, because [kargs](https://github.com/openshift/machine-config-operator/blob/b29dceb12638f3d7f30f339729344de69824a09b/pkg/daemon/update.go#L2237) and  [extensions](https://github.com/openshift/machine-config-operator/blob/b29dceb12638f3d7f30f339729344de69824a09b/pkg/daemon/update.go#L2242) were historically never gated by the diff, they ran every time: 

(now whether we liked that or not, I don't know, but I don't want to regress here)  

I can also stuff the condition in an "OsWorkNeedsToBeDone()" (or whatever) helper if that's clearer. 


I can stuff the condition in an "OsWorkNeedsToBeDone()" (or whatever) helper if that's clearer.

Reply:
Hmm, so as I understand it, this is really just making sure everything in the old https://github.com/openshift/machine-config-operator/blob/b36482885ba1304e122e7c01c26cd671dfdd0418/pkg/daemon/update.go#L332 update path is covered, and then this really is gating the new image based path more properly without affecting the legacy path, is that the correct way to interpret this?

Reply:
Right, with the old format image, the kernel and the extensions update ran every time we got to `applyOSChanges`, and they run `rpm-ostree update` which is totally fine with the old format images because all that did was update the packages if there were any. 

 I was trying to preserve that behavior. And if there weren't any extensions/kernels in would have effectively been a no-op. 

However, if we run `rpm-ostree update` on a new format image, it tries to check to see if there is an image update , and if we don't have a pull secret, we degrade, so that definitely can't run every time:
```
sh-4.4# rpm-ostree update
Pulling manifest: ostree-unverified-image:docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:f138db6071b64c4db6a48fb746970e7ee486c6b81b11ecc670f172a4767c7c41
error: remote error: reading manifest sha256:f138db6071b64c4db6a48fb746970e7ee486c6b81b11ecc670f172a4767c7c41 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
```

So yeah, for the sake of the native format images I was trying to make sure we only risk using the pull secret if we actually need to make a decision based on what type of image we have, while preserving as much of the legacy behavior as possible. 

We don't have a way to let the "legacy route" run every time exactly how it used to because we have to check the image type to know it's the legacy route without checking, and if we don't have a pull secret _that check will kill us_ :disappointed:

Reply:
Ack, thanks for the clarification!